src/rules/require_await.rs

```diff
@@ -0,0 +1,549 @@
+  Method(String),
+  #[display(fmt = "Async method has no 'await' expression.")]
+  AnonymousMethod,
+}
+
+#[derive(Display)]
+enum RequireAwaitHint {
+  #[display(
+    fmt = "Remove 'async' keyword from the function or use 'await' expression inside."
+  )]
+  RemoveOrUse,
+}
+
+impl LintRule for RequireAwait {
+  fn new() -> Box<Self> {
+    Box::new(RequireAwait)
+  }
+
+  fn tags(&self) -> &'static [&'static str] {
+    &["recommended"]
+  }
+
+  fn code(&self) -> &'static str {
+    CODE
+  }
+
+  fn lint_program(
+    &self,
+    context: &mut Context,
+    program: &swc_ecmascript::ast::Program,
+  ) {
+    let mut visitor = RequireAwaitVisitor::new(context);
+    program.visit_with(program, &mut visitor);
+  }
+
+  fn docs(&self) -> &'static str {
+    r#"Disallows async functions that have no await expression
+
+In general, the primary reason to use async functions is to use await expressions inside.
+If an async function has no await expression, it is most likely an unintentional mistake.
+
+### Invalid:
+```typescript
+async function f1() {
+  doSomething();
+}
+
+const f2 = async () => {
+  doSomething();
+};
+
+const f3 = async () => doSomething();
+
+const obj = {
+  async method() {
+    doSomething();
+  }
+};
+
+class MyClass {
+  async method() {
+    doSomething();
+  }
+}
+```
+
+### Valid:
+```typescript
+await asyncFunction();
+
+function normalFunction() {
+  doSomething();
+}
+
+async function f1() {
+  await asyncFunction();
+}
+
+const f2 = async () => {
+  await asyncFunction();
+};
+
+const f3 = async () => await asyncFunction();
+
+async function f4() {
+  for await (const num of asyncIterable) {
+    console.log(num);
+  }
+}
+
+// empty functions are valid
+async function emptyFunction() {}
+const emptyArrowFunction = async () => {};
+
+// generators are also valid
+async function* gen() {
+  console.log(42);
+}
+```
+"#
+  }
+}
+
+enum FunctionKind {
+  Function(Option<String>),
+  ArrowFunction,
+  Method(Option<String>),
+}
+
+impl From<FunctionKind> for RequireAwaitMessage {
+  fn from(kind: FunctionKind) -> Self {
+    match kind {
+      FunctionKind::Function(Some(name)) => RequireAwaitMessage::Function(name),
+      FunctionKind::Function(None) => RequireAwaitMessage::AnonymousFunction,
+      FunctionKind::ArrowFunction => RequireAwaitMessage::ArrowFunction,
+      FunctionKind::Method(Some(name)) => RequireAwaitMessage::Method(name),
+      FunctionKind::Method(None) => RequireAwaitMessage::AnonymousMethod,
+    }
+  }
+}
+
+impl Default for FunctionKind {
+  fn default() -> Self {
+    FunctionKind::Function(None)
+  }
+}
+
+struct FunctionInfo {
+  kind: FunctionKind,
+  is_async: bool,
+  is_generator: bool,
+  is_empty: bool,
+  has_await: bool,
+  upper: Option<Box<FunctionInfo>>,
+}
+
+#[derive(Default)]
+struct FunctionInfoBuilder {
+  kind: Option<FunctionKind>,
+  is_async: Option<bool>,
+  is_generator: Option<bool>,
+  is_empty: Option<bool>,
+  upper: Option<Box<FunctionInfo>>,
+}
```

Comment:
@magurotuna is this builder really required? Seems you could construct `FunctionInfo` in place where it's required - seems like a bunch of not really needed code